l1-contracts/src/UniFiAVSManager.sol

```diff
@@ -212,6 +232,169 @@ contract UniFiAVSManager is UniFiAVSManagerStorage, IUniFiAVSManager, UUPSUpgrad
+            bytes32 storedBlsPubKeyHash = $.validatorIndexes[validators[i].index];
+            if (storedBlsPubKeyHash != bytes32(0) && storedBlsPubKeyHash != validators[i].blsPubKeyHash) {
+                revert ValidatorIndexAlreadyUsed();
+            }
+
+            // Check if the registration salt has been used before
+            if (validatorRegistrationData.salt == validators[i].salt) {
+                revert SaltAlreadyUsed();
+            }
+
+            // Check if the signature is expired
+            if (block.timestamp > validators[i].expiry) {
+                revert SignatureExpired();
+            }
+
+            // Store the validator data
+            $.validators[validators[i].blsPubKeyHash] = ValidatorData({
+                eigenPod: address(0), // Not from EigenPod
+                index: validators[i].index, // Store the provided index
+                operator: msg.sender,
+                registeredUntil: type(uint64).max
+            });
+
+            bytes32 registrationHash = keccak256(
+                abi.encodePacked(
+                    validators[i].registrationSignature.X,
+                    validators[i].registrationSignature.Y,
+                    validators[i].expiry,
+                    validators[i].salt
+                )
+            );
+
+            $.validatorRegistrations[validators[i].blsPubKeyHash] = ValidatorRegistrationData({
+                registrationHash: registrationHash,
+                salt: validators[i].salt,
+                registeredAt: uint64(block.number),
+                activeAfter: uint64(block.number) + $.registerationDelay
+            });
+
+            $.validatorIndexes[validators[i].index] = validators[i].blsPubKeyHash;
+
+            emit ValidatorRegistered({
+                podOwner: address(0),
+                operator: msg.sender,
+                delegateKey: delegateKey,
+                blsPubKeyHash: validators[i].blsPubKeyHash,
+                validatorIndex: validators[i].index
+            });
+
+            newValidatorCount++;
+        }
+
+        // Update the operator's validator count
+        OperatorData storage operator = $.operators[msg.sender];
+        operator.validatorCount += uint128(newValidatorCount);
+    }
+
+    /**
+     * @inheritdoc IUniFiAVSManager
+     * @dev Restricted in this context is like `whenNotPaused` modifier from Pausable.sol
+     */
+    function verifyValidatorSignatures(ValidatorRegistrationSlashingParams[] calldata validators) external restricted {
+        UniFiAVSStorage storage $ = _getUniFiAVSManagerStorage();
+
+        for (uint256 i = 0; i < validators.length; i++) {
+            bytes32 blsPubKeyHash = BLSSignatureCheckerLib.hashG1Point(validators[i].pubkeyG1);
+
+            ValidatorData storage validator = $.validators[blsPubKeyHash];
+            ValidatorRegistrationData memory validatorRegistrationData = $.validatorRegistrations[blsPubKeyHash];
```

Comment:
This line is not necessary and it is making this function more expensive. You are copying the whole struct from storage -> memory, and on line 341 you are only using `registrationHash`. You can inline it on 341 `$.validatorRegistrations[blsPubKeyHash].registrationHash`